### PR TITLE
`@fast` attribute to enable early (0-cycle) transitions between alternating dynamic/static groups

### LINKED ISSUE #1828: Zero-cycle transitions from dynamic to static control
To slightly expand on this:

> Some of the grouping we were initially trying to have, a blocking group, a servicing group, an assert and deassert group, had to be combined together due to the time it takes to transfer between groups.

I actually think this is a super interesting problem that has implications for static/dynamic Calyx assignments. Artistically, we would like to be able to manage an AXI transaction with a control program like this:

```
while <more transactions to do> {
  seq {
    assert_ready;  // tell the remote side we are ready for their data
    wait_for_valid;  // wait for them to send us stuff
    par {
      do_transfer;  // actually do things the data on the bus
      deassert_ready;  // tell them to wait while we process the data
    }
  }
}
```

The problem is that AXI, of course, is super cycle-timing-sensitive. Remember that one of the core invariants is:

> On **any cycle** when both `ready=1` and `valid=1`, a transaction occurs.

So the remote side can schedule these transaction cycles back-to-back, i.e., 3 transactions can happen in exactly 3 cycles. Most saliently for us, say the remote side just keeps its `valid` signal high all the time and relies on us to assert/deassert `ready` to do transfers. Then, if we need to take some time to process a transfer, we have to be absolutely sure we only assert `ready` for one cycle at a time.

The problem with dynamic Calyx is that it offers no guarantees about how long it takes to go from `wait_for_valid` above to `deassert_ready`. And in practice, it actually does waste at least one cycle, meaning that the remote side sees us do two transfers when we only wanted one.

The big upshot of static Calyx is that it *does* offer these guarantees, so we would like to be able to just sprinkle the `static` everywhere get the timing we want. There is a critical problem, however: **the `wait_for_valid` group is fundamentally dynamic**. Our "dynamic islands" approach does not support sequencing a dynamic group with a static one with a zero-cycle transition latency.

So the big design question that I think this opens up is: is there some rule we can make up that would allow this kind of sequencing? That is, to allow some version of `seq { a ; b }` where `a` is dynamic and `b` is static where we can guarantee that `b` runs in the same cycle where `a` says it is done. It seems hard to do in general but maybe there is something we can do for restricted cases?

_Originally posted by @sampsyo in https://github.com/calyxir/calyx/issues/1820#issuecomment-1866308023_

### PR DESCRIPTION
**Pull request by ethanuppal**
Will eventually close #1828 once finalized and thoroughly tested (the former of which may take a while).

### CODE REVIEW

**Comment by ethanuppal** [comment]
I want to confirm the following: `@fast` in its most primitive state consists of applying the unstable early-transitions feature (which schedules 0-cycle transitions between groups `g1` and `g2` without regard to whether `g1[done] && g2[done]` is UNSAT) in an environment where we know it it is always possible (enforced via a check the `well-formed` pass).

File: tests/correctness/fast-attr/valid/fast6inc.expect [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (rachitnigam) [inline tests/correctness/fast-attr/valid/fast6inc.expect:1]: I think for this test specifically, we'd like to track the cycle count.

Comment (ethanuppal) [reply tests/correctness/fast-attr/valid/fast6inc.expect:1]: I was thinking about that. The issue I was running into is what if we modify the compilation process to add some other form of optimization that improves the cycle count further?

Comment (rachitnigam) [reply tests/correctness/fast-attr/valid/fast6inc.expect:1]: I think that is probably fine. This is also a different kind of correctness test which requires exact cycle count information (maybe like the `static` tests) so we should put it into a different folder where all tests always report their cycle counts?

Comment (ethanuppal) [reply tests/correctness/fast-attr/valid/fast6inc.expect:1]: I was also wondering if we could just do the fuzzing idea and guarantee the `\leq`? But that's not as strong. Ok, for now I'll add back in the cycle counts; of course, the very point of snapshot testing is that you can inevitably update your snapshots on significant changes.

File: runt.toml [hunk 2]
```diff
@@ -390,6 +388,32 @@ fud e --from systolic --to jq \
...
 expect_dir = "tests/correctness/systolic/mmult-expect"
 
+[[tests]]
+name = "[correctness] `@fast` attribute should not change program behavior"
+paths = ["tests/correctness/fast-attr/valid/*.futil"]
+cmd = """
+fud exec --from calyx --to jq \
+         --through dat \
+         --through icarus-verilog \
+         -s calyx.flags "-p validate -p compile -p lower" \
```
Comment (calebmkim) [inline runt.toml:398]: Just a drive-by comment: it might be worth disabling `static-promotion` on these tests; a promoted `static` seq will follow the `@fast` invariant, but `static` seqs get compiled before `tdcc`. Disabling `static-promotion` will make sure the tdcc changes you made are being tested.

**Comment by ethanuppal** [comment]
@rachitnigam @calebmkim update ping

**Comment by calebmkim** [comment]
I think this all lgtm!

**Comment by ethanuppal** [comment]
@rachitnigam update ping

File: calyx-opt/src/passes/top_down_compile_control.rs [hunk 3]
```diff
@@ -828,6 +834,7 @@ impl Schedule<'_, '_> {
             &if_stmt.tbranch,
             tru_transitions,
             early_transitions,
+            false,
```
Comment (rachitnigam) [inline calyx-opt/src/passes/top_down_compile_control.rs:837]: Why does this unconditionally pass `false` instead of threading through `early_transitions`? Is it because we want to enable this if we see it in a `seq` or something?

File: tools/cache/cache.futil [hunk 4]
```diff
@@ -0,0 +1,34 @@
+// This file was generated
```
Comment (rachitnigam) [inline tools/cache/cache.futil:1]: Do we need to commit this file?

**Review by rachitnigam — COMMENTED** [review]
Mostly looks good! One question about the implementation and one seemingly erroneously committed file. If the file is meant to be committed, then let's document what it is meant to be doing.

**Comment by rachitnigam** [comment]
@ethanuppal can you finalize the changes and answer the questions so we can merge this PR?

File: tools/cache/cache.futil (hunk 4 above)
Comment (ethanuppal) [reply tools/cache/cache.futil:1]: No

File: calyx-opt/src/passes/top_down_compile_control.rs (hunk 3 above)
Comment (ethanuppal) [reply calyx-opt/src/passes/top_down_compile_control.rs:837]: We only want to enable `@fast` on the parent control, I believe, unless you had different semantics in mind?

**Comment by ethanuppal** [comment]
Sorry for not getting around to this sooner!jk:wq

File: calyx-opt/src/passes/top_down_compile_control.rs (hunk 3 above)
Comment (rachitnigam) [reply calyx-opt/src/passes/top_down_compile_control.rs:837]: Okay, that makes sense! We should eventually get rid of the other `early-transition` logic entirely in a follow-up PR.

Comment (ethanuppal) [reply calyx-opt/src/passes/top_down_compile_control.rs:837]: Hmmm, there may be more cases in which we can apply `early-transition`, for example, if (as @sampsyo said) one could prove that two adjacent groups cannot be `done` at the same time; here, we could reuse `@fast`'s functionality, but there might be another case where it does apply recursively. That can be dealt with later though, it's not too hard to just re-add a parameter to all the functions.
